Transform Feedback: Add Support to Model.draw and minor API changes.

1chandu opened the pull request:
- Add TF support to Model class.
- Move `withParameters` from` Model.draw` to `Program.draw`, so temporary state settings can be applied in both case.
- Return early (no state push/pop) from `withParametes` when `parameters` object is empty.
  

Verified using test-browser, examples and deck.gl Wind example.

Pessimistress commented:
Update docs?

ibgreen commented on src/webgl/program.js:
```diff
@@ -86,19 +86,19 @@ export default class Program extends Resource {
-          this.gl.enable(GL.RASTERIZER_DISCARD);
-        }
+      if (parameters) {
+        log.error('Program.draw({prameters}) not needed to disable RASTERIZATION, \
```
I am not sure about this. 

It is not too hard to imagine cases when one wants to both draw and get transform feedback. Forcing it always off seems a bit too strict, we are taking something away from the user.

Shutting off the rasterizer by default could make sense though, if it is clearly documented. Just not forcing it off.

1chandu replied:
Agree, I will make this default behavior with a flag that disables it.

ibgreen commented on src/core/model.js:
```diff
@@ -329,7 +329,8 @@ export default class Model extends Object3D {
     settings,
-    framebuffer = null
+    framebuffer = null,
+    transformFeedback = null
```
Don't we also need to update the model constructor, to pass through varyings parameter to the Program constructor?

Pessimistress replied:
+1

1chandu commented:
@ibgreen @pessimistress
Addressed comments and updated the docs.

Regarding setting `RASTERIZED_DISCARD` automatically when `transformfeedback` is enabled, I am reverting this change since it complicates the API.  First, existing code is not using our state management, so setting and unsetting this state could change previous state. Second even if we offer to set this state in default case, what if user is also passing this state in `parameters`, which one should take precedence, we could decide on one but could be a confusing API, hence removing this logic.

I also moved `withParameters` from `model.draw` to `program.draw` so temporary state setting can done in both cases. In case of transform feedback, users can set RASTERIZER_DISCARD in parameters.

Pessimistress commented:
Verified with deck.gl attribute transition manager. One more thing about `TransformFeedback.bindBuffers`:

`bindBuffers(buffers = {}, {clear, varyingMap = {}})`
- The second argument should fallback to empty object
- `varyingMap` is not used nor documented

1chandu commented:
> varyingMap is not used nor documented
Yes, I would like to address this in a different PR. In TF,  buffer binding indices are decided on `varyings` array set on `program`.  For now the wind and attribute transitions demos working because they follow the same order, but what we want is able to specify either index or attribute name, when attribute names are specified we need to map them to correct index based on `varyings` array, this is not hooked up properly will address in a different PR.

ibgreen commented on src/webgl/context-state.js:
```diff
@@ -138,6 +138,18 @@ export function setParameters(gl, parameters) {
 export function withParameters(gl, parameters, func) {
   // assertWebGLContext(gl);
 
+  let emptyParametes = true;
```
+1 nice optimization

Typo: `emptyParameters`, or just `empty`?

Comment: `// check if parameters object is empty`

ibgreen commented on src/webgl/context-state.js:
```diff
@@ -138,6 +138,18 @@ export function setParameters(gl, parameters) {
+  }
+  /* eslint-enable no-unused-vars  */
+
+  if (emptyParametes) {
```
Comment: `// Avoid setting state if no parameters provided. Just call and return`

ibgreen commented on src/webgl/program.js:
```diff
@@ -94,39 +95,33 @@ export default class Program extends Resource {
       this.gl.useProgram(this.handle);
 
       if (transformFeedback) {
-        if (parameters[GL.RASTERIZER_DISCARD]) {
```
Maybe check if `undefined`, and in that case set it to true, otherwise as the user specified it.

If we don't use `withParameters`, we get the risk of an exception leaving the flag set/unset.

1chandu replied:
Since we are now calling `withParameters` inside, program.draw, any users specified setting will be taken care properly without polluting any previous state. As  commented above, not enforcing `rasterizer_discard` during TF flow, gives required flexibility in the API.

ibgreen commented on src/webgl/program.js:
```diff
@@ -94,39 +95,33 @@ export default class Program extends Resource {
-      } else {
-        this.gl.drawArrays(drawMode, offset, vertexCount);
-      }
+      withParameters(this.gl, parameters,
```
Slightly secondary, I believe we have a duplicate of this draw function in another file, we should remove it asap.

ibgreen commented on src/webgl/transform-feedback.js:
```diff
@@ -52,6 +52,7 @@ export default class TranformFeedback extends Resource {
   // program.use, should we move these methods (begin/pause/resume/end) to the Program?
   begin(primitiveMode) {
     this._bindBuffers();
+    this.gl.enable(GL.RASTERIZER_DISCARD);
```
This needs to be changed. 

Disabling rasterizer is the most common case, but not necessarily something the user always wants.

ibgreen replied:
I am also not 100% sure that other unrelated models couldn't be drawn between a begin and end, this setting would affect them.

1chandu replied:
I removed these, not needed now that users can specify this in `parameters` object.
